components/bl-cascadeSelect-component/src/helpers.js

```diff
@@ -0,0 +1,121 @@
+export const validate = (cascade, setItemsCascade, setParentItems, setItems) => {
+  const { isCircular, cycleLocation } = analyzeCircularDependencies(cascade);
+
+  if (isCircular) {
+    throw new Error('cascade have cycling object in ' + cycleLocation);
+  }
+
+  if (cascade) {
+    setItemsCascade(prepareCascade(cascade, setParentItems, setItems));
+  }
+}
+function analyzeCircularDependencies(obj) {
+  const keys = [];
+  const stack = [];
+  const stackSet = new Set();
+  let isCircular = false;
+  let cycleLocation;
+
+  function detect(obj, key) {
+    if (obj && typeof obj != 'object') {
+      return;
+    }
+
+    if (stackSet.has(obj)) {
+      cycleLocation = keys.join('.') + '.' + key;
+      isCircular = true;
+
+      return;
+    }
+
+    keys.push(key);
+    stack.push(obj);
+    stackSet.add(obj);
+
+    for (const k in obj) {
+      if (Object.prototype.hasOwnProperty.call(obj, k)) {
+        detect(obj[k], k);
+      }
+    }
+
+    keys.pop();
+    stack.pop();
+    stackSet.delete(obj);
+
+    return;
+  }
+
+  detect(obj, 'obj');
+
+  return { isCircular, cycleLocation };
+}
+
+const prepareCascade = (cascade, setParentItems, setItems) => {
+  let levelOfNesting = 0;
+  const parentItems = [];
+  const items = [];
+
+  const prepare = cascade => {
+    const validCascade = cascade.map(item => {
+      let validItem = { ...item, levelOfNesting };
+
+      if (item.children) {
+        levelOfNesting++;
+        validItem = {
+          ...validItem,
+          children: prepare(item.children),
+        };
+
+        parentItems.push({ code: item.code, isOpen: false, levelOfNesting });
+      } else {
+        items.push(validItem);
+      }
+
+      return validItem;
+    });
+
+    levelOfNesting--;
```

Comment:
could you please explain the logic around `++/-- levelOfNesting`

if it needs to calculate nesting depth would not it be better to pass a new level into the `prepare ` function?

```
 const prepare = (children, depth) => {
if (item.children) {
        validItem = {
          ...validItem,
          children: prepare(item.children, depth + 1),
        };

        parentItems.push({ code: item.code, isOpen: false, depth });
      } else {
        items.push(validItem);
      }
 }

 const preparedCascade = prepare(cascade);
```

Reply:
As far as I understand, with this implementation, we will not be able to get the last depth value when this recursion ends. It's needed to get max depth

Reply:
why do you need to know about the max depth?

Reply:
get sorted and separated array of `parentItems` by `levelOfNesting` in the `getNestedItems` function